go-ethereum-guides/write-basic-quiz-dapp-in-go/write-basic-quiz-dapp-in-go.md

```diff
@@ -0,0 +1,1069 @@
+    - `mapping (address=>bool) internal _leaderBoard`: 
+    Stores a hash map of user accounts and a boolean value 
+    that tells us whether a given account 
+    has answered the question correctly.
+    We've also set this state variable as `internal`
+    to prevent external callers from modifying its contents.
+2. The `constructor` is called when we deploy the contract.
+Here, we give it a question `_qn` and an answer `_ans`.
+    - We take `_qn` as a string because we mean for 
+    it to be easily readable by anyone interacting with the contract.
+    - Our answer `_ans` is set as a fixed slice of 32 bytes (`bytes32`)
+    because we want to store it as a `keccak256` hash. 
+    Hashing the value of `_ans` obscures it,
+    making it unreadable in the contract source or
+    the transaction logs.
+3. `sendAnswer()` allows us to send an answer to the contract.
+The answer sent to the contract must be a 32 byte
+keccak256 hash, which we compare to the value of 
+`_answer` stored on the contract.
+If the values match, we update our `leaderBoard` to show that
+the account that makes this function call has answered correctly.
+4. `_updateLeaderBoard()` takes a true/false value and sets
+the entry on the `_leaderBoard` mapping for our user's account to that value.
+It's an `internal` function, which prevents external callers
+from arbitarily modifying the `_leaderBoard` mapping.
+5. `checkBoard()` lets the user check the entry on the `_leaderBoard` mapping
+for their account, and whether they have been recorded as answering the question
+correctly. This function doesn't require EVM code execution as it's just
+returning the contents a state variable. It doesn't cost gas, and can be set 
+the `view` modifier.
+
+Now that we've got our Solidity contract fleshed out,
+we need to compile it to an ABI JSON specification and a contract binary.
+Then, we'll generate a Go binding file from those files, and import
+it into our Go DApp.
+
+We'll use `solc` and `abigen` to do this.
+
+First, compile your contract:
+
+1. Open the terminal.
+2. In your project directory, navigate to the "quiz" 
+    folder that contains your `quiz.sol` file.
+3. Run:
+
+    ```bash
+    solc --abi --bin quiz.sol -o build
+    ```
+
+This creates a "build" folder that contains the files `Quiz.abi` and `Quiz.bin`.
+
+Next, generate the Go binding file. In the "quiz" directory, run:
+
+```bash
+abigen --abi="build/Quiz.abi" --bin="build/Quiz.bin" --pkg=quiz --out="quiz.go"
+```
+
+This generates a Go file that contains bindings for your smart contract
+which we can import into our Go code.
+
+## Write Go Code
+
+### Connect to Rinkeby network and get account balance
+
+We'll start writing our Go DApp by initializing a connection to
+the Rinkeby network, using the Infura.io gateway endpoint that we 
+[set up earlier](#set-up-rinkeby-testnet-endpoint-on-infuraio).
+
+In your project directory,
+create a new `main.go` file and add the following code:
```

Comment:
@zeddee Two things here, First is that I assume this means the `quiz` directory? Second, it looks like a dependency is missing from the `go.mod` file - github.com/sqs/goreturns

Reply:
1. Nope, `main.go` should be in the `project directory root` (which yeah could be made clearer). So the dir structure (so far) should look like this:

```
./project/
-- /quiz/
---- /build/
------- Quiz.abi
------- Quiz.bin
---- quiz.sol
---- quiz.go
-- main.go
-- .env
-- go.mod
```

2. Is Go giving you an error for `github.com/sqs/goreturns`? I can't find it anywhere in my own code — plus I spun up a docker container + git cloned my repo to run it, and it's not there.

Reply:
@zeddee Odd, when I ran this outside of VSC it didn't ask for that dependency, maybe it's something to do with their integration.

Reply:
hmm yeah. +given that `go mod` is pretty new, and did not get released without community dissent. but it should work? i test my code with the official `golang:1.11` docker image, so that's the cleanest env i can get re: dev environments